File: cpp/src/heston_fourier.cpp
// heston_fourier.cpp
//
// See heston_fourier.hpp for design notes and theory/phase4/block2_heston_fourier.tex
// for the mathematical derivation.

#include "heston_fourier.hpp"

#include <array>
#include <cmath>
#include <stdexcept>

namespace quant::heston {

// =====================================================================
// Internal helpers (anonymous namespace)
// =====================================================================
namespace {

// pi -- portable constant rather than std::numbers::pi (C++20).
constexpr double PI = 3.14159265358979323846;

// 32-point Gauss-Legendre nodes on [-1, 1]
constexpr std::array<double, 32> GL32_NODES = {{
    -9.972638618494815699e-01, -9.856115115452683817e-01,
    -9.647622555875063899e-01, -9.349060759377396668e-01,
    -8.963211557660520912e-01, -8.493676137325699704e-01,
    -7.944837959679423856e-01, -7.321821187402897113e-01,
    -6.630442669302152314e-01, -5.877157572407623043e-01,
    -5.068999089322293594e-01, -4.213512761306353327e-01,
    -3.318686022821276671e-01, -2.392873622521370647e-01,
    -1.444719615827964876e-01, -4.830766568773832426e-02,
    +4.830766568773832426e-02, +1.444719615827964876e-01,
    +2.392873622521370647e-01, +3.318686022821276671e-01,
    +4.213512761306353327e-01, +5.068999089322293594e-01,
    +5.877157572407623043e-01, +6.630442669302152314e-01,
    +7.321821187402897113e-01, +7.944837959679423856e-01,
    +8.493676137325699704e-01, +8.963211557660520912e-01,
    +9.349060759377396668e-01, +9.647622555875063899e-01,
    +9.856115115452683817e-01, +9.972638618494815699e-01,
}};

// 32-point Gauss-Legendre weights
constexpr std::array<double, 32> GL32_WEIGHTS = {{
    +7.018610009470505756e-03, +1.627439473090574323e-02,
    +2.539206530926202410e-02, +3.427386291302176452e-02,
    +4.283589802222683568e-02, +5.099805926237609144e-02,
    +5.868409347853556501e-02, +6.582222277636168295e-02,
    +7.234579410884833806e-02, +7.819389578707022781e-02,
    +8.331192422694670696e-02, +8.765209300440378326e-02,
    +9.117387869576377979e-02, +9.384439908080451087e-02,
    +9.563872007927470831e-02, +9.654008851472765940e-02,
    +9.654008851472765940e-02, +9.563872007927470831e-02,
    +9.384439908080451087e-02, +9.117387869576377979e-02,
    +8.765209300440378326e-02, +8.331192422694670696e-02,
    +7.819389578707022781e-02, +7.234579410884833806e-02,
    +6.582222277636168295e-02, +5.868409347853556501e-02,
    +5.099805926237609144e-02, +4.283589802222683568e-02,
    +3.427386291302176452e-02, +2.539206530926202410e-02,
    +1.627439473090574323e-02, +7.018610009470505756e-03,
}};

// Integrate f on [a, b] via 32-point Gauss-Legendre.
//
// Maps [-1, 1] linearly to [a, b]:  x_i = mid + half_width * node_i,
// integral approximated by  half_width * sum_i w_i * f(x_i).
//
// Exact for polynomials of degree <= 63 on [a, b]; for our smooth,
// analytic integrand this gives near-double precision.
template <typename F>
double integrate_gl32(F&& f, double a, double b) {
    const double mid = 0.5 * (a + b);
    const double half_width = 0.5 * (b - a);
    double sum = 0.0;
    for (std::size_t i = 0; i < 32; ++i) {
        const double x = mid + half_width * GL32_NODES[i];
        sum += GL32_WEIGHTS[i] * f(x);
    }
    return half_width * sum;
}

// Standard normal CDF via std::erf:
//     N(x) = 0.5 * (1 + erf(x / sqrt(2))).
double standard_normal_cdf(double x) {
    return 0.5 * (1.0 + std::erf(x / std::sqrt(2.0)));
}

// Common pre-validation for pricing functions.
void check_basic(double tau, double S0, const HestonParams& p) {
    p.validate();
    if (tau <= 0.0) {
        throw std::invalid_argument("tau must be positive");
    }
    if (S0 <= 0.0) {
        throw std::invalid_argument("S0 must be positive");
    }
    // Note: r is unrestricted (negative rates are admissible).
}

}  // namespace


// =====================================================================
// HestonParams::validate
// =====================================================================

void HestonParams::validate() const {
    if (kappa <= 0.0) {
        throw std::invalid_argument("HestonParams: kappa must be positive");
    }
    if (theta <= 0.0) {
        throw std::invalid_argument("HestonParams: theta must be positive");
    }
    if (sigma <= 0.0) {
        throw std::invalid_argument("HestonParams: sigma must be positive");
    }
    if (rho < -1.0 || rho > 1.0) {
        throw std::invalid_argument("HestonParams: rho must be in [-1, 1]");
    }
    if (v0 < 0.0) {
        throw std::invalid_argument("HestonParams: v0 must be non-negative");
    }
}


// =====================================================================
// Characteristic function: AMSST formulation
// =====================================================================

std::pair<std::complex<double>, std::complex<double>>
heston_cf_coefficients(std::complex<double> u,
                        double tau,
                        double r,
                        const HestonParams& p) {
    using namespace std::complex_literals;

    // beta(u) = kappa - i rho sigma u
    const auto beta = p.kappa - 1i * p.rho * p.sigma * u;

    // d(u) = sqrt(beta^2 + sigma^2 u (u + i))
    // std::sqrt on std::complex returns the principal branch (Re d >= 0),
    // which is the correct branch choice for the AMSST formulation.
    const auto d = std::sqrt(beta * beta
                              + p.sigma * p.sigma * u * (u + 1i));

    // g(u) = (beta - d) / (beta + d)        [AMSST / "Little Trap"]
    const auto g = (beta - d) / (beta + d);

    const auto exp_neg_dtau = std::exp(-d * tau);
    const auto one_minus_g_exp = 1.0 - g * exp_neg_dtau;

    // D(tau; u) = (beta - d) / sigma^2 * (1 - exp(-d tau)) / (1 - g exp(-d tau))
    const auto D = (beta - d) / (p.sigma * p.sigma)
                   * (1.0 - exp_neg_dtau) / one_minus_g_exp;

    // C(tau; u) = i u r tau + (kappa theta / sigma^2)
    //              * [(beta - d) tau - 2 ln((1 - g exp(-d tau)) / (1 - g))]
    const auto log_term = std::log(one_minus_g_exp / (1.0 - g));
    const auto C = 1i * u * r * tau
                   + (p.kappa * p.theta / (p.sigma * p.sigma))
                     * ((beta - d) * tau - 2.0 * log_term);

    return {C, D};
}


std::complex<double>
heston_cf(std::complex<double> u,
          double tau,
          double S0,
          double r,
          const HestonParams& p) {
    using namespace std::complex_literals;
    check_basic(tau, S0, p);

    const auto [C, D] = heston_cf_coefficients(u, tau, r, p);
    return std::exp(C + D * p.v0 + 1i * u * std::log(S0));
}


// =====================================================================
// Lewis call price
// =====================================================================

double heston_call_lewis(double K,
                          double tau,
                          double S0,
                          double r,
                          const HestonParams& p,
                          double u_max) {
    using namespace std::complex_literals;
    check_basic(tau, S0, p);

    if (K <= 0.0) {
        throw std::invalid_argument("K must be positive");
    }
    if (u_max <= 0.0) {
        throw std::invalid_argument("u_max must be positive");
    }

    const double log_S_over_K = std::log(S0 / K);
    const double v0 = p.v0;

    // Lewis integrand (real-valued, smooth on [0, infty)):
    //
    //     Re[ exp(i u log(S0/K)) * phi_norm(u - i/2; tau) / (u^2 + 1/4) ].
    //
    // The variable u in the lambda corresponds to the real part of the
    // shifted Fourier variable; the imaginary shift -i/2 lives in the
    // argument of phi_norm.
    auto integrand = [&](double u) -> double {
        const std::complex<double> u_shifted{u, -0.5};
        const auto [C, D] = heston_cf_coefficients(u_shifted, tau, r, p);
        const auto phi_norm = std::exp(C + D * v0);
        const auto val = std::exp(1i * u * log_S_over_K) * phi_norm
                         / (u * u + 0.25);
        return val.real();
    };

    // Composite Gauss-Legendre on three panels [0, 5], [5, 30], [30, u_max].
    // The Lewis integrand decays as 1/u^2 modulated by a fast Gaussian-like
    // factor coming from the characteristic function. Most of its mass sits
    // in [0, 5] (where 1/(u^2 + 1/4) is largest), with a slow tail beyond.
    // Three asymmetric panels resolve the peak without wasting nodes on
    // the deep tail. If u_max is small enough that the splits collapse,
    // we fall back to a single-panel evaluation on [0, u_max].
    double I = 0.0;
    if (u_max <= 5.0) {
        I = integrate_gl32(integrand, 0.0, u_max);
    } else if (u_max <= 30.0) {
        I = integrate_gl32(integrand, 0.0, 5.0)
          + integrate_gl32(integrand, 5.0, u_max);
    } else {
        I = integrate_gl32(integrand, 0.0, 5.0)
          + integrate_gl32(integrand, 5.0, 30.0)
          + integrate_gl32(integrand, 30.0, u_max);
    }

    return S0 - std::sqrt(S0 * K) * std::exp(-r * tau) / PI * I;
}


// =====================================================================
// Reference: Black-Scholes and put-call parity
// =====================================================================

double black_scholes_call(double S0,
                           double K,
                           double tau,
                           double r,
                           double sigma) {
    if (tau <= 0.0) {
        return std::max(S0 - K, 0.0);
    }
    if (sigma <= 0.0) {
        return std::max(S0 - K * std::exp(-r * tau), 0.0);
    }

    const double sqrt_tau = std::sqrt(tau);
    const double d1 = (std::log(S0 / K) + (r + 0.5 * sigma * sigma) * tau)
                      / (sigma * sqrt_tau);
    const double d2 = d1 - sigma * sqrt_tau;

    return S0 * standard_normal_cdf(d1)
           - K * std::exp(-r * tau) * standard_normal_cdf(d2);
}


double put_via_parity(double call_price,
                      double S0,
                      double K,
                      double tau,
                      double r) {
    return call_price - S0 + K * std::exp(-r * tau);
}

}  // namespace quant::heston

File: python/quantlib/heston_fourier.py
"""
Heston model: characteristic function and Fourier-based pricing.

This module implements the Heston (1993) stochastic-volatility model under
the risk-neutral measure, with two pricing routes:

    1. Carr-Madan FFT (production method, used for surface pricing).
    2. Lewis adaptive quadrature (validation / single-strike high-precision).

The characteristic function uses the Albrecher-Mayer-Schoutens-Tistaert (2007)
"Little Trap" formulation, which avoids the branch-cut issue of the original
Heston (1993) formulation. See ``theory/phase4/block2_heston_fourier.tex``
for the derivation and discussion.

Notation
--------
The Heston dynamics under Q are

    dS_t = r S_t dt + sqrt(v_t) S_t dW1_t
    dv_t = kappa (theta - v_t) dt + sigma sqrt(v_t) dW2_t
    d<W1, W2>_t = rho dt

with parameters (kappa, theta, sigma, rho, v0) and risk-free rate r.
All pricing functions take (tau, S0, v0, r, kappa, theta, sigma, rho)
in that order.

References
----------
[Heston1993]  Heston, S. (1993). A closed-form solution for options with
              stochastic volatility. Review of Financial Studies 6(2).
[AMSST2007]   Albrecher, Mayer, Schoutens, Tistaert (2007). The little
              Heston trap. Wilmott Magazine.
[CarrMadan]   Carr, P., Madan, D. (1999). Option valuation using the FFT.
              Journal of Computational Finance 2(4).
[Lewis2000]   Lewis, A. (2000). Option Valuation under Stochastic Volatility.
              Finance Press.
"""

from __future__ import annotations

import numpy as np
from scipy.integrate import quad
from scipy.stats import norm


# =====================================================================
# Input validation
# =====================================================================

def _check_heston_params(tau, S0, v0, kappa, theta, sigma, rho):
    """Validate Heston model parameters. Raises ValueError on failure."""
    if tau <= 0:
        raise ValueError(f"tau must be positive, got {tau}")
    if S0 <= 0:
        raise ValueError(f"S0 must be positive, got {S0}")
    if v0 < 0:
        raise ValueError(f"v0 must be non-negative, got {v0}")
    if kappa <= 0:
        raise ValueError(f"kappa must be positive, got {kappa}")
    if theta <= 0:
        raise ValueError(f"theta must be positive, got {theta}")
    if sigma <= 0:
        raise ValueError(f"sigma must be positive, got {sigma}")
    if not (-1.0 <= rho <= 1.0):
        raise ValueError(f"rho must be in [-1, 1], got {rho}")


# =====================================================================
# Heston characteristic function (AMSST formulation)
# =====================================================================

def _heston_cf_coefficients(u, tau, r, kappa, theta, sigma, rho):
    """
    Compute the (C, D) coefficients of the Heston characteristic function.

    The characteristic function of X_T = log(S_T) under the risk-neutral
    measure, given (X_t = log S_t, v_t), takes the affine form

        phi(u; tau, X_t, v_t) = exp(C(tau; u) + D(tau; u) v_t + i u X_t).

    This helper returns the (C, D) pair so that callers can either build
    the full characteristic function (heston_cf below) or the normalized
    version exp(C + D v_t) used in Lewis's formula.
    """
    u = np.asarray(u, dtype=complex)

    beta = kappa - 1j * rho * sigma * u
    # Discriminant: beta^2 + sigma^2 (u^2 + i u) = beta^2 + sigma^2 u (u + i)
    d = np.sqrt(beta * beta + sigma * sigma * u * (u + 1j))

    g = (beta - d) / (beta + d)
    exp_neg_dtau = np.exp(-d * tau)

    one_minus_g_exp = 1.0 - g * exp_neg_dtau
    D = (beta - d) / (sigma * sigma) * (1.0 - exp_neg_dtau) / one_minus_g_exp

    log_term = np.log(one_minus_g_exp / (1.0 - g))
    C = (1j * u * r * tau
         + (kappa * theta / (sigma * sigma)) * ((beta - d) * tau - 2.0 * log_term))

    return C, D


def heston_cf(u, tau, S0, v0, r, kappa, theta, sigma, rho):
    """
    Heston characteristic function of log(S_T).

    Returns

        phi(u; tau, S0, v0) = E[ exp(i u log S_T) | S_t = S0, v_t = v0 ]
                            = exp(C + D v0 + i u log S0).

    Sanity checks satisfied by this formula (used as smoke tests):
        phi(0;  tau, ...) = 1
        phi(-i; tau, ...) = S0 * exp(r tau)            (=  E[S_T] under Q)
    """
    _check_heston_params(tau, S0, v0, kappa, theta, sigma, rho)

    C, D = _heston_cf_coefficients(u, tau, r, kappa, theta, sigma, rho)
    u_arr = np.asarray(u, dtype=complex)
    return np.exp(C + D * v0 + 1j * u_arr * np.log(S0))


# =====================================================================
# Carr-Madan FFT pricing
# =====================================================================

def heston_call_carr_madan(K, tau, S0, v0, r, kappa, theta, sigma, rho,
                            alpha=1.5, N=4096, eta=0.25):
    """
    Heston European call price(s) via Carr-Madan FFT.

    Following Carr & Madan (1999), the damped call price
    c(k) = exp(alpha k) C(k) is integrable for alpha large enough; its
    Fourier transform is

        psi(v) = exp(-r tau) phi(v - (alpha+1) i; tau)
                 / [(alpha + i v)(alpha + 1 + i v)],

    and the call price is recovered by

        C(k) = exp(-alpha k)/pi * integral_0^infty Re[exp(-i v k) psi(v)] dv.

    The FFT setup uses the Carr-Madan grid k_m = -b + m * lambda for
    m = 0, ..., N-1, with lambda = 2 pi / (N eta) and b = N lambda / 2.
    Strikes are interpolated linearly in log-strike to the requested K.
    """
    _check_heston_params(tau, S0, v0, kappa, theta, sigma, rho)

    if alpha <= 0:
        raise ValueError(f"alpha must be positive, got {alpha}")
    if N <= 0 or (N & (N - 1)):
        raise ValueError(f"N must be a positive power of 2, got {N}")
    if eta <= 0:
        raise ValueError(f"eta must be positive, got {eta}")

    K_array = np.atleast_1d(np.asarray(K, dtype=float))
    if np.any(K_array <= 0):
        raise ValueError("All strikes must be positive")

    # Carr-Madan natural strike grid (symmetric around log K = 0)
    lambd = 2.0 * np.pi / (N * eta)
    b = 0.5 * N * lambd
    k_grid = -b + np.arange(N) * lambd

    # Integration grid in v
    v = np.arange(N) * eta

    # Carr-Madan transform of the call: psi(v)
    u_complex = v - (alpha + 1.0) * 1j
    phi_at_u = heston_cf(u_complex, tau, S0, v0, r, kappa, theta, sigma, rho)
    denom = alpha * alpha + alpha - v * v + 1j * (2.0 * alpha + 1.0) * v
    psi = np.exp(-r * tau) * phi_at_u / denom

    # Simpson 1/3 weights, 0-indexed:
    #   w_0 = 1/3, w_n = 4/3 for n odd, w_n = 2/3 for n even (n >= 2).
    n_idx = np.arange(N)
    weights = np.where(n_idx == 0, 1.0 / 3.0,
                       (3.0 - (-1.0) ** n_idx) / 3.0)

    # FFT input: integration step, phase factor and weights folded in
    x = eta * np.exp(1j * v * b) * psi * weights
    fft_out = np.fft.fft(x)

    # Recover call prices on the natural grid k_m
    C_grid = np.real(np.exp(-alpha * k_grid) * fft_out / np.pi)

    # Linear interpolation in log-strike to the requested K values
    log_K_query = np.log(K_array)
    if np.any(log_K_query < k_grid[0]) or np.any(log_K_query > k_grid[-1]):
        raise ValueError(
            "Some requested strikes fall outside the FFT grid; "
            "increase N or adjust eta."
        )
    C_query = np.interp(log_K_query, k_grid, C_grid)

    if np.ndim(K) == 0:
        return float(C_query[0])
    return C_query


# =====================================================================
# Lewis adaptive quadrature pricing
# =====================================================================

def heston_call_lewis(K, tau, S0, v0, r, kappa, theta, sigma, rho,
                      u_max=200.0, quad_kwargs=None):
    """
    Heston European call price(s) via Lewis adaptive quadrature.

    Uses Lewis's (2000) inversion formula in the form

        C(K, T) = S0 - sqrt(K S0) exp(-r tau) / pi
                  * integral_0^infty Re[ exp(i u log(S0/K))
                                         * phi_norm(u - i/2; tau)
                                         / (u^2 + 1/4) ] du,

    where phi_norm(u; tau) = exp(C(tau; u) + D(tau; u) v_0) is the Heston
    characteristic function with X_t = 0 (i.e., S_t = 1). This is
    equivalent to the standard Lewis formula but expressed in terms of
    phi_norm rather than the centered char function of log(F_T/F_0); see
    block2_heston_fourier.tex for the derivation.

    For single-strike pricing this typically beats Carr-Madan in both
    accuracy and speed; for many strikes, Carr-Madan FFT wins.
    """
    _check_heston_params(tau, S0, v0, kappa, theta, sigma, rho)

    if u_max <= 0:
        raise ValueError(f"u_max must be positive, got {u_max}")

    if quad_kwargs is None:
        quad_kwargs = {}

    K_array = np.atleast_1d(np.asarray(K, dtype=float))
    if np.any(K_array <= 0):
        raise ValueError("All strikes must be positive")

    prices = np.empty(len(K_array), dtype=float)
    log_S0 = np.log(S0)

    for i, Ki in enumerate(K_array):
        log_S_over_K = log_S0 - np.log(Ki)

        def integrand(u):
            C_co, D_co = _heston_cf_coefficients(
                u - 0.5j, tau, r, kappa, theta, sigma, rho
            )
            phi_norm = np.exp(C_co + D_co * v0)
            integrand_complex = (np.exp(1j * u * log_S_over_K) * phi_norm
                                 / (u * u + 0.25))
            return float(np.real(integrand_complex))

        integral, _ = quad(integrand, 0.0, u_max, **quad_kwargs)
        prices[i] = (S0 - np.sqrt(S0 * Ki) * np.exp(-r * tau) / np.pi
                     * integral)

    if np.ndim(K) == 0:
        return float(prices[0])
    return prices


# =====================================================================
# Reference: Black-Scholes for sanity checks
# =====================================================================

def black_scholes_call(S0, K, tau, r, sigma):
    """
    Black-Scholes European call price.

    Used as a reference benchmark in the BS limit. When sigma_Heston -> 0
    in the Heston model, the variance v(t) becomes deterministic and
    integrated variance V(T) = theta T + (v_0 - theta)(1 - exp(-kappa T))/kappa.
    The Heston call price approaches the BS call with effective volatility
    sqrt(V(T)/T); in the special case v_0 = theta this is just sqrt(v_0).
    """
    if tau <= 0:
        return max(S0 - K, 0.0)
    if sigma <= 0:
        return max(S0 - K * np.exp(-r * tau), 0.0)

    sqrt_tau = np.sqrt(tau)
    d1 = ((np.log(S0 / K) + (r + 0.5 * sigma * sigma) * tau)
          / (sigma * sqrt_tau))
    d2 = d1 - sigma * sqrt_tau
    return S0 * norm.cdf(d1) - K * np.exp(-r * tau) * norm.cdf(d2)


def put_via_parity(call_price, S0, K, tau, r):
    """European put price via put-call parity: P = C - S0 + K exp(-r tau)."""
    return call_price - S0 + K * np.exp(-r * tau)
